src/busctl/busctl.c

```diff
@@ -2281,6 +2281,108 @@ static int get_property(int argc, char **argv, void *userdata) {
+
+                        r = sd_bus_message_dump(msg, stdout, 0);
+                        if (r < 0)
+                                return r;
+                } else {
+
+                        fputs(sd_bus_message_get_signature(msg, true), stdout);
+                        fputc(' ', stdout);
+
+                        r = format_cmdline(msg, stdout, false);
+                        if (r < 0)
+                                return bus_log_parse_error(r);
+
+                        fputc('\n', stdout);
+                }
+        }
+
+        return 0;
+}
+
+static int wait_signal(int argc, char **argv, void *userdata) {
+        _cleanup_(sd_bus_flush_close_unrefp) sd_bus *bus = NULL;
+        _cleanup_(sd_bus_message_unrefp) sd_bus_message *reply = NULL;
+        _cleanup_(sd_bus_error_free) sd_bus_error error = SD_BUS_ERROR_NULL;
+        bool fini = false;
+        int argn = 1;
+        int r;
+
+        const char *sender = argc == 5 ? argv[argn++] : NULL;
+        const char *path = argv[argn++];
+        const char *interface = argv[argn++];
+        const char *member = argv[argn++];
+
+        r = acquire_bus(/* set_monitor= */ false, &bus);
+        if (r < 0)
+                return r;
+
+        if (sender && !service_name_is_valid(sender))
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Invalid service name: %s", sender);
+        if (!object_path_is_valid(path))
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Invalid object path: %s", path);
+        if (!interface_name_is_valid(interface))
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Invalid interface name: %s", interface);
+        if (!member_name_is_valid(member))
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Invalid member name: %s", member);
+
+        r = sd_bus_match_signal(bus, NULL, sender, path, interface, member, on_bus_signal, &fini);
+        if (r < 0)
+                return log_error_errno(r, "Failed to match signal %s on interface %s: %m", member, interface);
+
+        /* The match is installed and we're ready to observe the signal */
+        sd_notify(/* unset_environment= */ false, "READY=1");
+
+        for (;;) {
+                r = sd_bus_process(bus, NULL);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to process bus: %m");
+                if (r > 0)
+                        continue;
+
+                if (fini)
+                        break;
+
+                r = sd_bus_wait(bus, arg_timeout > 0 ? arg_timeout : UINT64_MAX);
+                if (r == 0 && arg_timeout > 0) {
+                        if (!arg_quiet && arg_json_format_flags == SD_JSON_FORMAT_OFF)
+                                log_info("Timed out waiting for messages, exiting.");
+                        return 0;
+                }
+                if (r < 0)
+                        return log_error_errno(r, "Failed to wait for bus: %m");
+        }
```

Comment:
Not bad, but the loop can be simplified with a timer event source for timeout and sd_event_loop().
Then, `fini` can be dropped and call `sd_event_exit()` in the handler.

Reply:
Okay, I changed it to use sd_event. PTAL.